src/k2/module/k23si/Module.cpp

```diff
@@ -835,90 +918,117 @@ bool K23SIPartitionModule::_makeProjection(dto::SKVRecord::Storage& fullRec, dto
+        .then([] {
+            return seastar::make_ready_future<Status>(dto::K23SIStatus::Created("TR created"));
         })
-        .handle_exception_type([this](TxnManager::ClientError&) {
+        .handle_exception_type([this](TxnManager::ClientError& err) {
+            // Failed to create
+            K2LOG_D(log::skvsvr, "Partition: {}, failed creating TR: {}", _partition, err.what());
+            return seastar::make_ready_future<Status>(dto::K23SIStatus::AbortConflict("txn too old in write"));
+        })
+        .handle_exception_type([this](TxnManager::ServerError& err) {
             // Failed to create
-            K2LOG_D(log::skvsvr, "Partition: {}, failed creating TR", _partition);
-            return RPCResponse(dto::K23SIStatus::AbortConflict("txn too old in write"), dto::K23SIWriteResponse{});
+            K2LOG_D(log::skvsvr, "Partition: {}, failed creating TR, {}", _partition, err.what());
+            return seastar::make_ready_future<Status>(dto::K23SIStatus::InternalError("Unable to create transaction"));
         });
+}
+
+seastar::future<std::tuple<Status, dto::K23SIWriteResponse>>
+K23SIPartitionModule::handleWrite(dto::K23SIWriteRequest&& request, FastDeadline deadline) {
+    if (!_validateRequestPartition(request)) {
+        // tell client their collection partition is gone
+        return RPCResponse(dto::K23SIStatus::RefreshCollection("collection refresh needed in write"), dto::K23SIWriteResponse());
     }
+    // NB: failures in processing a write do not require that we set the TR state to aborted at the TRH. We rely on
+    //     the client to do the correct thing and issue an abort on a failure.
+    K2LOG_D(log::skvsvr, "Partition: {}, handle write: {}", _partition, request);
+    if (request.designateTRH) {
+        K2LOG_D(log::skvsvr, "Partition: {}, designating TRH in {}", _partition, request.mtr);
+        return _designateTRH({.trh=request.trh, .mtr=request.mtr})
+            .then([this, request=std::move(request), deadline] (auto&& status) mutable {
+                if (!status.is2xxOK()) {
+                    K2LOG_D(log::skvsvr, "Partition: {}, failed creating TR for {}", _partition, request.mtr);
+                    return _respondAfterFlush(std::move(status), dto::K23SIWriteResponse{});
+                }
 
-    auto& versions = _indexer[request.key];
-    // in this situation, return AbortRequestTooOld error.
-    {
-        Status validateStatus = _validateStaleWrite(request, versions);
-        if (!validateStatus.is2xxOK()) {
-            K2LOG_D(log::skvsvr, "Partition: {}, request too old for key {} due to {}",
-                    _partition, request.key, validateStatus);
-            return RPCResponse(std::move(validateStatus), dto::K23SIWriteResponse{});
+                K2LOG_D(log::skvsvr, "Partition: {}, succeeded creating TR. Processing write for {}", _partition, request.mtr);
+                request.designateTRH = false;
```

Comment:
Why do we unset the flag? Is it needed at all?

Reply:
No, it isn't needed. It was there before so I kept it in, but we can remove